### PR TITLE
[Feature] Tracking Information Requests

### PR DESCRIPTION
**Pull request by mbabker**
Pull Request for Issue #9

### Summary of Changes

This will be the central piece of the user workflow for managing information requests.  Not entirely "complete", but functional enough to introduce a PR and share the methodology and approach here.

### Testing Instructions

1) Download this branch
2) Install Joomla
3) Navigate to `Components > Privacy`
4) Interact with component

### Documentation Changes Required

Absolutely 😉

#### General Notes

- Purposefully this system will NOT support deleting requests (if this really becomes burdensome we can revisit later but this screen can act as an audit log and we shouldn't allow that data to be dropped easily)

#### Screenshots

Main list view
screen shot 2018-05-06 at 1 13 54 pm

Add request screen
screen shot 2018-05-06 at 1 12 02 pm

Show request screen
screen shot 2018-05-06 at 1 14 16 pm

#### Requests Workflow

This PR implements one aspect of the data request workflow, the ability for the site administrator to log a request.  As it is likely a request will come from someone who does not have an account on the site, the identifier for these requests will be the user's email address; optionally this request MAY be linked to a user account.

Requests use a workflow style model for statuses which looks something like this:

- New Item
    - To: Pending
- Request Validated
    - From: Pending
    - To: Confirmed
- Request Completed
    - From: Confirmed
    - To: Completed
- Request Invalidated
    - From: Pending, Confirmed
    - To: Invalid

All requests start in a pending status.  GDPR requires identity verification for these types of requests, so requests in a pending status indicates that this verification has not yet been completed (for now this will be in the form of an email verification similar to password resets, but this isn't implemented yet).  A request can transition to either Confirmed or Invalid from here; Confirmed would indicate the user responding to the verification email and Invalid would indicate that the request timed out (TODO - Have a system plugin or CLI script that can be cron'ed that will check for pending requests older than X hours, presumably something like 24, 48, or 72, and invalidate them).  In a Confirmed status, either it can be acted upon (Completed) or the site admin can invalidate the request.  As indicated in the above list, a request in an Invalid or Completed status cannot be transitioned.

### TODO (not necessarily blockers to the merge of this PR but definitely work to be continued)

- [x] Disallow creating a request for the currently authenticated user
- [x] Disallow creating a new request for a user who has an active request (status of Pending or Confirmed), we should nudge admins to work with these active requests
- [ ] Think about a Failed status in addition to Invalid
    - This would be the "user never confirmed validation" status and could potentially give admins the possibility to retry validation if it expires (up to a sane limit)
- [ ] Think about some kind of help text directly on the single request layouts
    - This is a bit of an odd case, requests are basically immutable once received outside the status field so there may need to be more steering of users as to why this section works this way and what to do if they do make a mistake, something for some UX folks to consider
- [x] Support multiple request types (this can just be a new `request_type` column for now supporting "export" and "removal" types, leave this flexible enough to add more types if need be)
- [x] Implement post save hook for new items to send out the verification email

### CODE REVIEW

**Comment by sandewt** [comment]
Great, I have been able to reproduce the sreenshots. 

Missing link in componts to `administrator/index.php?option=com_privacy` yet.

**Comment by mbabker** [comment]
Unless I screwed something up in https://github.com/joomla-projects/privacy-framework/commit/4b0f1c1bbb314bf6d9377f8187f9549d4158de9f all the menu scaffolding should've been added (either way that's outside scope of this PR, this commit was made to the repo to have the component skeleton in place to make PRs against).

File: administrator/components/com_privacy/models/forms/filter_requests.xml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+	<fields name="list">
+		<field
+			name="fullordering"
+			type="list"
+			label="JGLOBAL_SORT_BY"
+			description="JGLOBAL_SORT_BY"
+			onchange="this.form.submit();"
+			default="a.id DESC"
+		>
```
Comment (brianteeman) [inline administrator/components/com_privacy/models/forms/filter_requests.xml]: indent one tab

**Comment by mbabker** [comment]
OK, aside from needing the post-create email hook and probably some UX love fundamentally this is ready for testing/review.

**Comment by brianteeman** [comment]
Seems to do what you intended it to do without error

**Comment by brianteeman** [comment]
I can't find a reference for this but my gut feeling is that you say
Requested at 2:30
Requested on 1st July
Requested on 1st July at 2:30

Requested at 2018-01-01 09:30 just seems wrong

but its a minor thing

**Comment by mbabker** [comment]
Grammatically speaking we'd need a `DATE_FORMAT_LC7="j F Y \a\t H:i:s"` string to do it right (turns into "1 December 2017 at 19:30:00").  Lacking that, what's there now seems to be the best option.

**Comment by brianteeman** [comment]
Or avoid the problem completely and change the field from "Requested at" to "Date Requested"

**Comment by mbabker** [comment]
Label changed.

**Comment by mbabker** [comment]
Added in the hook for sending the confirmation email out.  I'm done for this PR, at this point everything to start work on frontend elements is in place, there's now a TODO in code that needs us merging the branch with https://github.com/joomla/joomla-cms/pull/16879 merged over here.

**Comment by mbabker** [comment]
I'm content with where this is at.  Merging to allow next phase to be built out.

**Pull request merged**